### DISCUSSION TITLE
[@exp] OAuth "non interactive" testing

### DISCUSSION THREAD

**Discussion by laurentpayot** [firebase/firebase-js-sdk#4746] [Q&A] [answered]
[The OAuth "non interactive" testing method from the docs](https://firebase.google.com/docs/emulator-suite/connect_auth#non-interactive_testing_3) is not working with firebase@exp because [`OAuthProvider.credential()`](https://modularfirebase.web.app/reference/auth.oauthprovider.credential) does not take the same arguments as the old non-modular one. I tried passing the old-way token via the `accessToken` option without success. 

All my OAuth end-to-end tests (with playwright) are broken. Is it related to #4715?

**Comment by Feiyang1** [comment, marked as answer]
@sam-gc

**Reply by sam-gc** [reply]
Hi @laurentpayot, would you mind sharing a small snippet detailing how you're calling this?

The OAuth non-interactive testing method should work with `GoogleAuthProvider.credential()` ([ref](https://firebase.google.com/docs/reference/js/v9/auth.googleauthprovider.md#googleauthprovidercredential)) (or any other OAuth provider). The base class you mention (`OAuthProvider.credential()`) might work differently. I'll double check a bit later

**Reply by laurentpayot** [reply]
Hi @sam-gc,

I should have mentioned that I’m trying to have a generic `signInWithProvider(providerId: string)` function for all possible providers.
As you said `GoogleAuthProvider.credential()` is working with emulators in the snippet below but it is not provider-generic. Sadly `OAuthProvider.credentialFromJSON()` is crashing. 

```typescript
import { auth } from './authInit'
import {
  signInWithRedirect, OAuthProvider, browserPopupRedirectResolver,
  signInWithCredential, GoogleAuthProvider
} from '@firebase/auth'


import type { UserCredential } from '@firebase/auth'


export function signInWithProvider(providerId: string): Promise<UserCredential> {
  const provider = new OAuthProvider(providerId)
  if (location.hostname === 'localhost') {
    const token = JSON.stringify({ sub: 'abc123', email: 'oauthtestuser@gmail.com', email_verified: true })
    // const credential = OAuthProvider.credentialFromJSON(token) // not working (firebase error)
    // const credential = OAuthProvider.credential(token) // not working (no credential method)
    // const credential = provider.credential(token) // not working (bad argument type)
    const credential = GoogleAuthProvider.credential(token) // working but not provider-generic
    return signInWithCredential(auth, credential)
  } else {
    provider.addScope('profile')
    provider.addScope('email')
    return signInWithRedirect(auth, provider, browserPopupRedirectResolver)
  }
}
```

**Reply by sam-gc** [reply]
Ah, I understand now. The signature changed slightly for the generic OAuth credential method. Whereas before it could be a string OR an object, we now only accept an object. `credentialFromJSON` accepts both a string and an object, but all it does in the case of string is call `JSON.parse` and then go through the other method (see [this file](https://github.com/firebase/firebase-js-sdk/blob/5ad33abe4c1987417b3df38a78de3e2ac1799d04/packages-exp/auth-exp/src/core/providers/oauth.ts#L135)).

In the old SDK, if the first parameter was a string, it was treated as the ID token. In the new version, you need to specify this explicitly. Try:
```ts
const credential = provider.credential({ idToken: token });
```

Thanks for the detailed example!

**Reply by laurentpayot** [reply]
Awesome. I’m wondering why I didn’t try this. Thanks @sam-gc!
